### PR TITLE
feat: support get porperty to get nginx variables

### PR DESCRIPTION
**Pull request by bzp2010**
Support get Nginx variables by GetProperty. #53

### CODE REVIEW

**Comment by juzhiyuan** [comment]
@membphis Could we invite @bzp2010 to this repo?

**Comment by bzp2010** [comment]
## Update
I have some questions about the specific implementation.
Do we need to use Get Property to get other variables later (we can specify an implementation prefix, such as "ngx_var" or "var", for the call to get the Nginx Variable).

**Comment by spacewander** [comment]
> ## Update
> 
> I have some questions about the specific implementation. Do we need to use Get Property to get other variables later (we can specify an implementation prefix, such as "ngx_var" or "var", for the call to get the Nginx Variable).

IMHO, we don't need it, as Envoy doesn't add a `envoy_` prefix to its property. And `ngx_remote_addr` will be strange.

File: src/http/ngx_http_wasm_api.c [hunk 1]
```diff
@@ -311,7 +311,20 @@ proxy_get_property(int32_t path_data, int32_t path_size,
                                           res_data, res_size);
     }
 
-    return PROXY_RESULT_OK;
+    ngx_http_request_t *r;
+    must_get_req(r);
+    
+    u_char *lowcase_buf = ngx_palloc(r->pool, path_size);
```
Comment (spacewander) [inline src/http/ngx_http_wasm_api.c:317]: Could you follow Nginx's code style? You can take a look at other code. And we need to check the returned res.

File: t/testdata/property/main.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+type httpLifecycle struct {
+	// Embed the default http context here,
+	// so that we don't need to reimplement all the methods.
+	types.DefaultHttpContext
+	contextID uint32
+}
+
+var variables = []string{"arg_test", "scheme", "host", "request_uri", "uri"}
```
Comment (spacewander) [inline t/testdata/property/main.go:45]: Better to pass the test data via configuration. You can take a look at https://github.com/api7/wasm-nginx-module/pull/61

File: t/property.t [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+
+
+=== TEST2: get_property
+--- config
+location /t {
+    content_by_lua_block {
+        local wasm = require("resty.proxy-wasm")
+        local plugin = wasm.load("plugin", "t/testdata/property/main.go.wasm")
+        local plugin_ctx, err = wasm.on_configure(plugin, "test")
```
Comment (spacewander) [inline t/property.t:24]: Let's add a case which var is not found

File: t/property.t [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+run_tests();
+
+__DATA__
+
+=== TEST 1: load
+--- config
+location /t {
+    content_by_lua_block {
+        local wasm = require("resty.proxy-wasm")
+        assert(wasm.load("plugin", "t/testdata/property/main.go.wasm"))
```
Comment (spacewander) [inline t/property.t:12]: We don't need to test loading in the property test file.

**Review by spacewander — COMMENTED** [review]
Don't forget to update https://github.com/api7/wasm-nginx-module/blob/main/proxy_wasm_abi.md

File: t/property.t (hunk 4 above)
Comment (bzp2010) [reply t/property.t]: removed

File: t/testdata/property/main.go (hunk 2 above)
Comment (bzp2010) [reply t/testdata/property/main.go:45]: I migrated it to a table driven test implemented using test scripts.

File: src/http/ngx_http_wasm_api.c (hunk 1 above)
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:317]: I'm sorry I made some low mistakes. I've just contacted nginx C development. I'll read the document and improve it.

File: t/property.t (hunk 3 above)
Comment (bzp2010) [reply t/property.t]: I added the missing callback when the variable does not exist, and also added test cases for it. Now, when the user requests a non-existent variable, the proxy-wasm-sdk will return an error.

File: src/http/ngx_http_wasm_api.c (hunk 1 above)
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:317]: I performed the following code style adjustments follow this document. [https://nginx.org/en/docs/dev/development_guide.html#code_style](https://nginx.org/en/docs/dev/development_guide.html#code_style)

- Move variables that are not predefined in the function header to there
- Adjust the indentation of the function definition area to conform to the specification
- Split lines with a single line length of more than 80 characters according to the specification

Also, I have added a check for when the Nginx variable does not exist, which will return `NOT_FOUND` directly to wasm.

If I still have any missed code style issues, please help point them out and I will fix those immediately.

**Comment by membphis** [comment]
> @membphis Could we invite @bzp2010 to this repo?

current way is enough. this repo is an open source project

**Comment by juzhiyuan** [comment]
> > @membphis Could we invite @bzp2010 to this repo?
> 
> current way is enough. this repo is an open source project

What's the relationship with Open Source project and Invitation?

**Comment by juzhiyuan** [comment]
> > @membphis Could we invite @bzp2010 to this repo?
> 
> current way is enough. this repo is an open source project

What's the relationship with Open Source project and Invitation?

File: src/http/ngx_http_wasm_api.c [hunk 5]
```diff
@@ -310,8 +315,26 @@ proxy_get_property(int32_t path_data, int32_t path_size,
...
-
-    return PROXY_RESULT_OK;
+    
+    /*
+     * assemblyscript need plugin_root_id to initialize, and the state 
+     * will set after plugin initialization completed
+     */
+    must_get_req(r);
+    
+    lowcase_buf = ngx_palloc(r->pool, path_size);
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c:325]: get it

File: proxy_wasm_abi.md [hunk 6]
```diff
@@ -242,6 +242,15 @@ Add key (`key_data`, `key_size`) with the value (`value_data`, `value_size`) to
...
+### `proxy_get_property`
+
+* params:
+  - `i32 (const char*) path_data`
+  - `i32 (size_t) path_size`
+  - `i32 (const char*) res_data`
+  - `i32 (size_t) res_size`
+
+Get data such as Nginx variables and plugin ID, Use path_data as key and write the value of the obtained variable to res_data.
 ## HTTP (L7) extensions
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply proxy_wasm_abi.md:255]: added

File: src/http/ngx_http_wasm_api.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (bzp2010) [reply src/http/ngx_http_wasm_api.c]: changed

File: src/http/ngx_http_wasm_api.c [hunk 7]
```diff
@@ -310,8 +315,26 @@ proxy_get_property(int32_t path_data, int32_t path_size,
...
-    return PROXY_RESULT_OK;
+    
+    /*
+     * assemblyscript need plugin_root_id to initialize, and the state 
+     * will set after plugin initialization completed
+     */
+    must_get_req(r);
+    
+    lowcase_buf = ngx_http_wasm_get_string_buf(r->pool, path_size);
+    hash = ngx_hash_strlow(lowcase_buf, p, path_size);
```
Comment (spacewander) [inline src/http/ngx_http_wasm_api.c:326]:
```suggestion
    hash = ngx_hash_strlow(lowcase_buf, (u_char *) p, path_size);
```

**Review by spacewander — APPROVED** [review]

**Pull request merged**